pczt/src/orchard.rs

```diff
@@ -0,0 +1,535 @@
+            // Destructure `rhs` to ensure we handle everything.
+            let Action {
+                cv_net,
+                spend:
+                    Spend {
+                        nullifier,
+                        rk,
+                        spend_auth_sig,
+                        recipient,
+                        value,
+                        rho,
+                        rseed,
+                        fvk,
+                        witness,
+                        alpha,
+                        zip32_derivation: spend_zip32_derivation,
+                        dummy_sk,
+                        proprietary: spend_proprietary,
+                    },
+                output:
+                    Output {
+                        cmx,
+                        ephemeral_key,
+                        enc_ciphertext,
+                        out_ciphertext,
+                        recipient: output_recipient,
+                        value: output_value,
+                        rseed: output_rseed,
+                        ock,
+                        zip32_derivation: output_zip32_derivation,
+                        proprietary: output_proprietary,
+                    },
+                rcv,
+            } = rhs;
+
+            if lhs.cv_net != cv_net
+                || lhs.spend.nullifier != nullifier
+                || lhs.spend.rk != rk
+                || lhs.output.cmx != cmx
+                || lhs.output.ephemeral_key != ephemeral_key
+                || lhs.output.enc_ciphertext != enc_ciphertext
+                || lhs.output.out_ciphertext != out_ciphertext
+            {
+                return None;
+            }
+
+            if !(merge_optional(&mut lhs.spend.spend_auth_sig, spend_auth_sig)
+                && merge_optional(&mut lhs.spend.recipient, recipient)
+                && merge_optional(&mut lhs.spend.value, value)
+                && merge_optional(&mut lhs.spend.rho, rho)
+                && merge_optional(&mut lhs.spend.rseed, rseed)
+                && merge_optional(&mut lhs.spend.fvk, fvk)
+                && merge_optional(&mut lhs.spend.witness, witness)
+                && merge_optional(&mut lhs.spend.alpha, alpha)
+                && merge_optional(&mut lhs.spend.zip32_derivation, spend_zip32_derivation)
+                && merge_optional(&mut lhs.spend.dummy_sk, dummy_sk)
+                && merge_map(&mut lhs.spend.proprietary, spend_proprietary)
+                && merge_optional(&mut lhs.output.recipient, output_recipient)
+                && merge_optional(&mut lhs.output.value, output_value)
+                && merge_optional(&mut lhs.output.rseed, output_rseed)
+                && merge_optional(&mut lhs.output.ock, ock)
+                && merge_optional(&mut lhs.output.zip32_derivation, output_zip32_derivation)
+                && merge_map(&mut lhs.output.proprietary, output_proprietary)
+                && merge_optional(&mut lhs.rcv, rcv))
```

Comment:
None of the effecting fields (here, the fields directly represented in the transaction) can be merged in a way that modifies them if we have any signatures.